### PR TITLE
add adpativeEN visualization and Psychrometric visualization

### PR DESCRIPTION
**Pull request by jiadongZhangWorkIn**
I have drawn two EN graphs and now need some comments

### CODE REVIEW

File: components/PsychrometricCharts.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
+import math
+import numpy as np
+import plotly.graph_objs as go
+from dash import Dash, dcc, html
+from dash.dependencies import Input, Output
+
+# 计算函数
+def calculate_saturation_vapor_pressure(temp_c):
+    return 610.94 * np.exp(17.625 * temp_c / (temp_c + 243.04))
```
Comment (FedericoTartarini) [inline components/PsychrometricCharts.py:9]: this is good but you should use the function from pythermalcomfort so we do not have code repetitions

File: components/PsychrometricCharts.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+def calculate_dew_point(temp_c, relative_humidity):
```
Comment (FedericoTartarini) [inline components/PsychrometricCharts.py:11]: this is good but you should use the function from pythermalcomfort so we do not have code repetitions

File: components/PsychrometricCharts.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    saturation_vapor_pressure = calculate_saturation_vapor_pressure(temp_c)
+    actual_vapor_pressure = relative_humidity / 100.0 * saturation_vapor_pressure
+    dew_point = 243.04 * np.log(actual_vapor_pressure / 610.94) / (17.625 - np.log(actual_vapor_pressure / 610.94))
+    return dew_point
+
+def calculate_wet_bulb_temp(dry_bulb_temp, relative_humidity):
```
Comment (FedericoTartarini) [inline components/PsychrometricCharts.py:17]: why do you need to calculate the wet bulb temperature for this chart?

File: components/PsychrometricCharts.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        wet_bulb_temp = dry_bulb_temp * math.atan(0.151977 * math.sqrt(relative_humidity + 8.313659)) \
+                        + math.atan(dry_bulb_temp + relative_humidity) \
+                        - math.atan(relative_humidity - 1.676331) \
+                        + 0.00391838 * (relative_humidity ** 1.5) * math.atan(0.023101 * relative_humidity) \
+                        - 4.686035
+        if abs(prev_wet_bulb_temp - wet_bulb_temp) < epsilon:
+            break
+    return wet_bulb_temp
+
+def calculate_absolute_humidity(dry_bulb_temp, relative_humidity):
```
Comment (FedericoTartarini) [inline components/PsychrometricCharts.py:31]: this is good but you should use the function from pythermalcomfort so we do not have code repetitions

File: components/PsychrometricCharts.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+    vapor_pressure_min = 0  
+    vapor_pressure_max = 3000
+
+    traces.append(go.Scatter(
+        x=np.concatenate([np.array([green_zone_1_left, green_zone_1_right]), 
+                          np.array([green_zone_1_right, green_zone_1_left])]),
+        y=np.concatenate([np.array([vapor_pressure_max, vapor_pressure_max]), 
+                          np.array([vapor_pressure_min, vapor_pressure_min])]),
+        fill='toself',
```
Comment (FedericoTartarini) [inline components/PsychrometricCharts.py:73]: reformat the code with Black

File: components/PsychrometricCharts.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        fillcolor='rgba(0, 255, 0, 0.6)',  
+        line=dict(color='rgba(255,255,255,0)'),
+        showlegend=False
+    ))
+
+    
+    for RH in relative_humidities:
```
Comment (FedericoTartarini) [inline components/PsychrometricCharts.py:102]: use PEP naming convention, RH should be in lower case

File: components/adaptiveEN.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    x_values = np.linspace(10, 30,5) 
+    
+    y_values_cat_i_up = []
+    y_values_cat_i_low = []
+    y_values_cat_ii_up = []
+    y_values_cat_ii_low = []
+    y_values_cat_iii_up = []
+    y_values_cat_iii_low = []
+
+    for x in x_values:
```
Comment (FedericoTartarini) [inline components/adaptiveEN.py:16]: this is not correct, you should not calculalate the the y values for all the x. You just need two x and two y values to draw a line, this is simple math. No need to loop through so many x values. It makes the code slower and the line does not look like a line

**Review by FedericoTartarini — COMMENTED** [review]
Please address my comments

File: components/adaptiveEN.py (hunk 7 above)
Comment (jiadongZhangWorkIn) [reply components/adaptiveEN.py:16]: I think this code only needs two points to draw a straight line. It should be that xy satisfies a linear equation. All I need to do is change "x_values ​​= np.linspace(10, 30, 5)" to "x_values ​​= np.array([10, 30])" in the code. Is there anything wrong with this AdaptiveEN graph?

File: components/PsychrometricCharts.py (hunk 1 above)
Comment (black6pepper) [reply components/PsychrometricCharts.py:9]: got it,sir i was working on that and almost done

File: components/PsychrometricCharts.py (hunk 5 above)
Comment (black6pepper) [reply components/PsychrometricCharts.py:73]: got it ,sir

File: components/PsychrometricCharts.py (hunk 6 above)
Comment (black6pepper) [reply components/PsychrometricCharts.py:102]: got it , i already make them same name as the prototype

File: components/adaptiveEN.py (hunk 7 above)
Comment (jiadongZhangWorkIn) [reply components/adaptiveEN.py:16]: image
here's the change i made,sir

**Comment by FedericoTartarini** [comment]
@jiadongZhangWorkIn could you please address my comments or close the pull request?

File: components/PsychrometricCharts.py (hunk 1 above)
Comment (jiadongZhangWorkIn) [reply components/PsychrometricCharts.py:9]: image
sir,i changed the code and used the function you mention before

File: components/PsychrometricCharts.py (hunk 2 above)
Comment (jiadongZhangWorkIn) [reply components/PsychrometricCharts.py:11]: image
after that i changed the calculation of the parameters of the chart

File: components/PsychrometricCharts.py (hunk 3 above)
Comment (jiadongZhangWorkIn) [reply components/PsychrometricCharts.py:17]: image
the calculation is now different from before after i delete the wet bulb temperature calculation

File: components/PsychrometricCharts.py (hunk 5 above)
Comment (jiadongZhangWorkIn) [reply components/PsychrometricCharts.py:73]: image
i already black the code,sir

File: components/PsychrometricCharts.py (hunk 6 above)
Comment (jiadongZhangWorkIn) [reply components/PsychrometricCharts.py:102]: image
i already make RH in lower case

File: components/PsychrometricCharts.py (hunk 4 above)
Comment (jiadongZhangWorkIn) [reply components/PsychrometricCharts.py:31]: image
the calculation is now different from before

**Pull request closed without merging**